enhancements/workload-partitioning/wide-availability-workload-partitioning.md

```diff
@@ -0,0 +1,731 @@
+}
+
+// PartitioningMode defines the CPU partitioning mode of the nodes.
+type CPUPartitioningMode string
+
+const (
+  // No partitioning is applied.
+	CPUPartitioningNone        CPUPartitioningMode = "None"
+  // All nodes are configured for partitioning.
+	CPUPartitioningAllNodes    CPUPartitioningMode = "AllNodes"
+)
+```
+
+### Openshift Installer
+
+We want to be able to allow a concise way to turn this feature on and easily
+enable other consumers of the openshift-installer to utilize it via the
+`InstallConfig`. This allows our other supported installer methods, such as
+assisted installer or ztp, a straightforward way to expose this feature.
+
+In order to set the correct `Infrastructure Status` at install time, we'll
+modify the `InstallConfig` to include enums to correctly set the status.
+
+```go
+type InstallConfig struct {
+	...
+	// CPUPartitioning determines if a cluster should be setup for CPU workload partitioning at install time.
+	// When this field is set the cluster will be flagged for CPU Partitioning allowing users to segregate workloads to
+	// specific CPU Sets. This does not make any decisions on workloads it only configures the nodes for it.
+	//
+	// This feature will alter the infrastructure nodes and prepare them for CPU Partitioning and as such can not be changed after being set.
+	//
+	// +kubebuilder:default="None"
+	// +optional
+	CPUPartitioning CPUPartitioningMode `json:"cpuPartitioningMode,omitempty"`
+  ...
+}
+
+
+// CPUPartitioningMode defines how the nodes should be setup for partitioning the CPU Sets.
+// +kubebuilder:validation:Enum=None;AllNodes
+type CPUPartitioningMode string
+
+const (
+	// CPUPartitioningNone means that no CPU Partitioning is on in this cluster infrastructure
+	CPUPartitioningNone CPUPartitioningMode = "None"
+	// CPUPartitioningAllNodes means that all nodes are configured with CPU Partitioning in this cluster
+	CPUPartitioningAllNodes CPUPartitioningMode = "AllNodes"
+)
+```
+
+### Machine Config Operator
+
+Once we have a global flag, and a way to set it at install time, we'll need to
+apply the needed configurations at install time during bootstrap. We will add
+this ability to MCO to generate and maintain the needed configurations before
+`kubelet` and the `api-server` stands up.
+
+We will add to the `kubelet` controller the ability to watch the
+`Infrastructure` resource and if CPU Partitioning is set to `AllNodes` we will
+generate the bootstrap and the controller will maintain the MCs from that point
+on. Things to note, this feature is explicitly designed to not be turned off, as
+such once set we will not remove the MCs.
+
+We will need to support upgrades for Single Node since this feature already
+exists for them but this implementation differs slightly. To avoid needless
+restarts, we will not alter the current configuration if we detect that we are
+in a single node topology and the nodes are already configured for CPU Partitioning.
```

Comment:
This one is tricky, I will need to update the wording to make it clear. The thing I'm calling out here is only the initial upgrade event. At install time what MCO will do is essentially prime the cluster for CPU Partitioning, during an upgrade this would mean that MCO will try to generate that file that was supposed to originally be there, that has the risk of overriding existing MCO configs that a user might have, and we will need to update their PerformanceProfile to include the new enum to recreate their config and get it back to their original state. That would mean 1 or possibly 2 extra restarts for them to get them back to their expected state. The idea with this is that for those consumers MCO doesn't try to do that if it's in a state where it existed prior, and as for NTO, if the user updates their PerformanceProfile to include the new enum, they will then initiate a restart as NTO will then take ownership from that point on. That has the added benefit of being customer driven, and predictable. I am completely open to discussing other approaches for this scenario.

One thing to note Ian, from this point on, the intention is for the PerformanceProfile (NTO) to then completely drive the CPUPartitioning so the user never touches MC files, and their PerformanceProfile defines the state of their CPUPartitioning for the cluster.

Reply:
The way that MCO will tell if we're already in a partitioned state for SNO is to check single node status and node capacity/allocatable (the current way) as well as if the expected bootstrap configs are present. So the logic boils down to

```go
if Topology == SNO && NodeCapacity HAS "workload/cores" && bootStrapConfig != Exists {
       skip
}
```

Reply:
sounds good

Reply:
We actually have customers that modified the cpu partitioning (disabled it) via Tuned override. That is a very unusual corner case, however PAO so far did not write any important cpu set information by itself (always via NTO and tuned). This design changes that and might make overrides more complicated. I am not sure the groups of users using SNO/WP and those overrides have any intersection though.

Reply:
That's interesting @MarSik are those customers using a `PerformanceProfile` combined with `Tuned` overrides?

Reply:
Given the new change with performance profile API no longer being changed, but deriving the crio and kubelet config from the `reserved` and `isolated` the current behavior described here shouldn't be effected since the only extra files being generated will be the crio config and everything else will behave as it currently is. In regard to crio overrides, we can support changing that for specific customers, by helping them create a file name in higher lexical order to override the file generated by NTO for crio. https://github.com/cri-o/cri-o/blob/main/docs/crio.conf.d.5.md#configuration-precedence